src/style/style.js

```diff
@@ -352,16 +351,22 @@ class Style extends Evented {
         } else {
             delete this.stylesheet.projection;
         }
-        this.updateProjection();
+        if (!this.map._explicitProjection) {
+            this.map._updateProjection(projection);
+        }
```

Comment:
Similar comment as below (line 360-364), I think this could be compressed for an early return if moved within `map._updateProjection` so that you always call `map._updateProjection` from here.